### PR TITLE
Cache endpoint availability

### PR DESCRIPTION
**Pull request by karenzheung**
Added fhir_endpoints_availability table that takes number of http response 200 returned and number of total responses for each endpoint
Add availability column to fhir_endpoints_info
Add migrate up scripts that will add column and table and populate from existing db
Add migrate down script to remove changes

To test (I recommend pgAdmin for this) - 
1. Run and populate the database
2. Check that fhir_endpoints_availability exist and is correctly populated
3. Check that fhir_endpoints_info has availability column. Availability should be 1.0 for endpoints that have 200 status
4. Check that fhir_endpoints_info_history also has this info
5. Change the http_response of endpoints and verify that availability is updated accordingly 

To test migrate scripts: 

1. run and populate the database on master
2. checkout this branch
3. run `make migrate_database`   - to migrate up
4. Use pgadmin to verify that availability table and column were created and populated
5. Change line 42 in `db/migration/cmd/main.go` to be `m.Steps(-2)`
6. run `make migrate_database`  - to migrate down
7. Verify that tables are gone 

Pull requests into this repository require the following checks to be completed by the submitter and reviewers.

**Submitter:**
- [x] Administrative tasks are complete.
  - This pull request describes why these changes were made.
  - The JIRA ticket links to this PR.
  - The JIRA ticket for this PR is: https://oncprojectracking.healthit.gov/support/browse/LANTERN-255
- [x] Tests are complete.
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.
- [x] Code review has been performed.
  - Code diff has been done and checked to ensure only expected code is being committed.
  - A linter for the language has been run.

**Primary Reviewer:**

- [ ] Code works.
  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
- [ ] Code review has been performed.
  - Code is maintainable, reusable, efficient, and correct.
  - Code accomplishes the tasks purpose.
  - Code follows style guidance appropriate for the language, including passing linter checks.
  - Code is well commented (which does not mean verbosely commented).
- [ ] Tests are complete.
  - Tests are included and test edge cases.
  - Tests have been run locally and pass.

**Secondary Reviewer:**

- [x] Code works.
  - Code has been run locally and the basic functionality affected by the PR has been checked.
  - Edge cases have been checked to ensure behavior is as expected.
- [x] Code review has been performed.
  - Code accomplishes the tasks purpose.
- [x] Tests are complete.
  - Tests have been run locally and pass.

**Merger:**

Prior to merging, ensure that all checks have been completed. Delete the branch after the merge
is complete.

Create a new branch `<branch name>_gomodupdate`. Run `make update_mods` to update the `go.mod`
and `go.sum` files to point to master. Create a PR. Require one sanity check reviewer.

### CODE REVIEW

File: scripts/sql/000003_add_endpoint_availability.up.sql [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        i             record;
+        okay_count      bigint;
+        all_count       bigint;
+    BEGIN
+        FOR i IN SELECT DISTINCT fhir_endpoints_info_history.url FROM fhir_endpoints_info_history
+        LOOP
+            SELECT COUNT(*) INTO all_count FROM fhir_endpoints_info_history WHERE url = i.url AND (operation = 'I' OR operation = 'U');
+            SELECT COUNT(*) INTO okay_count FROM fhir_endpoints_info_history WHERE url = i.url AND http_response = 200 AND (operation = 'I' OR operation = 'U');
+            INSERT INTO fhir_endpoints_availability VALUES (i.url, okay_count, all_count);
+            UPDATE fhir_endpoints_info SET availability = (okay_count * 1.0) / all_count WHERE url = i.url;
```
Comment (bmath10) [inline scripts/sql/000003_add_endpoint_availability.up.sql:30]: why is `okay_count` multiplied by `1.0` here?

File: scripts/sql/dbsetup.sql [hunk 2]
```diff
@@ -156,6 +192,14 @@ CREATE TABLE endpoint_organization (
     CONSTRAINT endpoint_org PRIMARY KEY (url, organization_npi_id)
 );
 
+CREATE TABLE fhir_endpoints_availability (
+    url             VARCHAR(500),
```
Comment (bmath10) [inline scripts/sql/dbsetup.sql:196]: do we always want to keep the availability, even if the the url no longer exists? my gut reaction is to delete the entry if the url is no longer in the `fhir_endpoints` table, but i can understand keeping it around to reference for historical purposes

File: scripts/sql/dbsetup.sql [hunk 3]
```diff
@@ -26,6 +26,40 @@ CREATE OR REPLACE FUNCTION add_fhir_endpoint_info_history() RETURNS TRIGGER AS $
...
+                RETURN NEW;
+            ELSE
+                INSERT INTO fhir_endpoints_availability VALUES (NEW.url, 0, 1);
+                NEW.availability = 0.00;
+                RETURN NEW;
+            END IF;
+        ELSE
+            IF NEW.http_response = 200 THEN
+                UPDATE fhir_endpoints_availability SET http_200_count = okay_count + 1, http_all_count = all_count + 1 WHERE url = NEW.url;
+                NEW.availability := (okay_count + 1.0) / (all_count + 1);
```
Comment (bmath10) [inline scripts/sql/dbsetup.sql:52]: I think all of these should either be `1` or `1.0` for consistency

File: scripts/sql/dbsetup.sql [hunk 4]
```diff
@@ -26,6 +26,40 @@ CREATE OR REPLACE FUNCTION add_fhir_endpoint_info_history() RETURNS TRIGGER AS $
...
+                RETURN NEW;
+            ELSE
+                UPDATE fhir_endpoints_availability SET http_200_count = okay_count, http_all_count = all_count + 1 WHERE url = NEW.url;
```
Comment (bmath10) [inline scripts/sql/dbsetup.sql:55]: I think you can just update the `http_all_count`, since the `http_200_count` isn't changing

File: scripts/sql/000003_add_endpoint_availability.up.sql (hunk 1 above)
Comment (karenzheung) [reply scripts/sql/000003_add_endpoint_availability.up.sql:30]: in postgres the result of the division is based on the type of the operands. Since okay_count and all_count are both ints I multiplied one by 1.0 to get a decimal result

**Comment by bmath10** [comment]
I had no issues with the migration stuff, but when I run this branch after running make clean, I'm getting a lot of warnings saying `Availability is not a float`. Have you seen this before?

File: scripts/sql/000003_add_endpoint_availability.up.sql (hunk 1 above)
Comment (bmath10) [reply scripts/sql/000003_add_endpoint_availability.up.sql:30]: oh okay that makes sense, thanks for the clarification!

File: scripts/sql/dbsetup.sql (hunk 2 above)
Comment (karenzheung) [reply scripts/sql/dbsetup.sql:196]: I asked Matt sometime ago about the scenario where if a endpoint was removed and added back again, should the availability of it include http statuses from before it was removed. And he had said yes if it's the same URL. So for that reason I didn't remove entries for removed URLs. But I could see the table getting really big if an endpoint was retired and not going to be in use again. @mayerm94 Thoughts?

**Comment by karenzheung** [comment]
> I had no issues with the migration stuff, but when I run this branch after running make clean, I'm getting a lot of warnings saying `Availability is not a float`. Have you seen this before?

The `Availability is not a float` warnings were coming from formatMessage. formatMessage doesn't actually need to check the availability field because that field will be filled in by a trigger. I had put it there because one of the test in Test_formatMessage wasn't passing. So now I removed the check from formatMessage and modified the test

File: scripts/sql/dbsetup.sql (hunk 2 above)
Comment (bmath10) [reply scripts/sql/dbsetup.sql:196]: i mean that makes sense to me!

**Comment by bmath10** [comment]
Looks good! I think it makes sense to wait to merge it in until the chpl_criteria_met PR gets merged in with the migrations stuff, but once that's in I'll approve this!
